File: src/Rect.ts
import Point2D from "./Point";
import Segment from "./Segment";
import GeneralMath from "GeneralMath";

class GRect{
	private constructor(
		private readonly p1: Point2D,
		private readonly p2: Point2D,
		private readonly p3: Point2D,
		private readonly p4: Point2D,
	) {}

	static createRectFromPoints(p1: Point2D, p2: Point2D, p3: Point2D, p4: Point2D): GRect {
		const centroidPoint = GeneralMath.centroid([p1, p2, p3, p4]);
		const points = [p1, p2, p3, p4];
		points.sort((a, b) => {
			return GeneralMath.getAngle_2Points(a, centroidPoint) > GeneralMath.getAngle_2Points(b, centroidPoint);
		});
		return new GRect(points[0], points[1], points[2], points[3]);
	}

	static createRectFromPart(part: { Size: Vector3; CFrame: CFrame }): GRect {
		const size = part.Size;
		const cframe = part.CFrame;
		const p1 = cframe.mul(new CFrame(-size.X / 2, 0, -size.Z / 2));
		const p2 = cframe.mul(new CFrame(size.X / 2, 0, -size.Z / 2));
		const p3 = cframe.mul(new CFrame(size.X / 2, 0, size.Z / 2));
		const p4 = cframe.mul(new CFrame(-size.X / 2, 0, size.Z / 2));
		return GRect.createRectFromPoints(
			new Point2D(p1.Position.X, p1.Position.Z),
			new Point2D(p2.Position.X, p2.Position.Z),
			new Point2D(p3.Position.X, p3.Position.Z),
			new Point2D(p4.Position.X, p4.Position.Z),
		);
	}

	getMaxPoint(): Point2D {
		const points = this.getPoints();
		const x = points.map((p) => p.getX());
		const y = points.map((p) => p.getY());
		const maxX = math.max(...x);
		const maxY = math.max(...y);
		return new Point2D(maxX, maxY);
	}

	getMinPoint(): Point2D {
		const points = this.getPoints();
		const x = points.map((p) => p.getX());
		const y = points.map((p) => p.getY());
		const minX = math.min(...x);
		const minY = math.min(...y);
		return new Point2D(minX, minY);
	}

	getSides(): [Segment, Segment, Segment, Segment] {
		return [
			new Segment(this.p1, this.p2),
			new Segment(this.p2, this.p3),
			new Segment(this.p3, this.p4),
			new Segment(this.p4, this.p1),
		];
	}

	getPoints(): [Point2D, Point2D, Point2D, Point2D] {
		return [this.p1, this.p2, this.p3, this.p4];
	}

	getSize(): Point2D {
		const length = GeneralMath.getDistance_2Points(this.p1, this.p2);
		const width = GeneralMath.getDistance_2Points(this.p2, this.p3);
		return new Point2D(length, width);
	}

	getArea(): number {
		const size = this.getSize();
		return size.getX() * size.getY();
	}

	getPerimeter(): number {
		const size = this.getSize();
		return 2 * (size.getX() + size.getY());
	}

	getRotation(): number {
		const points = this.getPoints();
		const dy = points[1].getY() - points[0].getY();
		const dx = points[1].getX() - points[0].getX();
		const angle = math.atan2(dy, dx);
		return angle;
	}

	getAxisAlignedRect(): GRect {
		const size = this.getSize();
		return GRect.createRectFromPoints(
			new Point2D(size.getX() / 2, size.getY() / 2),
			new Point2D(-size.getX() / 2, size.getY() / 2),
			new Point2D(-size.getX() / 2, -size.getY() / 2),
			new Point2D(size.getX() / 2, -size.getY() / 2),
		);
	}

	getBoundingBox(): GRect {
		const points = this.getPoints();
		const x = points.map((p) => p.getX());
		const y = points.map((p) => p.getY());
		const minX = math.min(...x);
		const maxX = math.max(...x);
		const minY = math.min(...y);
		const maxY = math.max(...y);
		return new GRect(
			new Point2D(minX, minY),
			new Point2D(maxX, minY),
			new Point2D(maxX, maxY),
			new Point2D(minX, maxY),
		);
	}

	isPointInside(point: Point2D): boolean {
		const [x, y] = GeneralMath.rotatePoint(point.getX(), point.getY(), -this.getRotation());
		const rotatedPoint = new Point2D(x, y);
		const axisAlignedRect = this.getAxisAlignedRect();
		const points = axisAlignedRect.getPoints();
		const xValues = points.map((p) => p.getX());
		const yValues = points.map((p) => p.getY());
		const minX = math.min(...xValues);
		const maxX = math.max(...xValues);
		const minY = math.min(...yValues);
		const maxY = math.max(...yValues);
		return rotatedPoint.getX() >= minX && rotatedPoint.getX() <= maxX && rotatedPoint.getY() >= minY && rotatedPoint.getY() <= maxY;
	}

	getCenter(): Point2D {
		return GeneralMath.centroid(this.getPoints());
	}

	intersects(segment: Segment): boolean {
		return this.getSides().some((side) => side.intersects(segment));
	}

	intersectsRect(rect: GRect): boolean {
		return this.getSides().some((side) => rect.intersects(side));
	}

	getIntersection(segment: Segment): Point2D | Segment | undefined {
		// Loop through each side of the rectangle and check for intersection
		const intersections = this.getSides().map((side: Segment) => side.getIntersection(segment));
		const nonUndefinedIntersections = new Array<Segment | Point2D>();
		for (const intersection of intersections) {
			if (intersection !== undefined) {
				nonUndefinedIntersections.push(intersection);
			}
		}

		// If the segment intersects, and the intersection is a segment, return the segment
		const segmentIntersections = nonUndefinedIntersections.filter(
			(intersection) => intersection instanceof Segment,
		);
		if (segmentIntersections.size() > 0) {
			return segmentIntersections[0];
		}

		// If the segment intersects, and the intersection is a point at the end of the segment, return the point
		const pointIntersections = nonUndefinedIntersections.filter(
			(intersection) => intersection instanceof Point2D,
		) as Point2D[];
		// If no intersections are found, return undefined
		// Because if there are no segment intersections, there will always be at least one point intersection
		if (pointIntersections.size() === 0) return undefined;

		// If the point intersections is only one, return the point
		if (pointIntersections.size() === 1) return pointIntersections[0];

		const intersectionAtRectCorner = this.getPoints().find((point) => {
			return pointIntersections.some((intersection) => {
				if (point.equals(intersection)) {
					return true;
				}
			});
		});
		if (intersectionAtRectCorner !== undefined) {
			return intersectionAtRectCorner;
		}

		// Else, find the next intersection, and return a segment from the first intersection to the next intersection
		const firstIntersection = pointIntersections[0];
		const secondIntersection = pointIntersections[1];
		return new Segment(firstIntersection, secondIntersection);
	}
}

export default GRect;

File: src/Rect.lua
-- Compiled with roblox-ts v2.3.0
local TS = require(game:GetService("ReplicatedStorage"):WaitForChild("rbxts_include"):WaitForChild("RuntimeLib"))
local Point2D = TS.import(script, game:GetService("ReplicatedStorage"), "TS", "Point").default
local Segment = TS.import(script, game:GetService("ReplicatedStorage"), "TS", "Segment").default
local GeneralMath = TS.import(script, game:GetService("ReplicatedStorage"), "TS", "GeneralMath").default
local GRect
do
	GRect = setmetatable({}, {
		__tostring = function()
			return "GRect"
		end,
	})
	GRect.__index = GRect
	function GRect.new(...)
		local self = setmetatable({}, GRect)
		return self:constructor(...) or self
	end
	function GRect:constructor(p1, p2, p3, p4)
		self.p1 = p1
		self.p2 = p2
		self.p3 = p3
		self.p4 = p4
	end
	function GRect:createRectFromPoints(p1, p2, p3, p4)
		local centroidPoint = GeneralMath.centroid({ p1, p2, p3, p4 })
		local points = { p1, p2, p3, p4 }
		table.sort(points, function(a, b)
			return GeneralMath.getAngle_2Points(a, centroidPoint) > GeneralMath.getAngle_2Points(b, centroidPoint)
		end)
		return GRect.new(points[1], points[2], points[3], points[4])
	end
	function GRect:createRectFromPart(part)
		local size = part.Size
		local cframe = part.CFrame
		local _cFrame = CFrame.new(-size.X / 2, 0, -size.Z / 2)
		local p1 = cframe * _cFrame
		local _cFrame_1 = CFrame.new(size.X / 2, 0, -size.Z / 2)
		local p2 = cframe * _cFrame_1
		local _cFrame_2 = CFrame.new(size.X / 2, 0, size.Z / 2)
		local p3 = cframe * _cFrame_2
		local _cFrame_3 = CFrame.new(-size.X / 2, 0, size.Z / 2)
		local p4 = cframe * _cFrame_3
		return GRect:createRectFromPoints(Point2D.new(p1.Position.X, p1.Position.Z), Point2D.new(p2.Position.X, p2.Position.Z), Point2D.new(p3.Position.X, p3.Position.Z), Point2D.new(p4.Position.X, p4.Position.Z))
	end
	function GRect:getMaxPoint()
		local points = self:getPoints()
		-- ▼ ReadonlyArray.map ▼
		local _newValue = table.create(#points)
		local _callback = function(p)
			return p:getX()
		end
		for _k, _v in points do
			_newValue[_k] = _callback(_v, _k - 1, points)
		end
		-- ▲ ReadonlyArray.map ▲
		local x = _newValue
		-- ▼ ReadonlyArray.map ▼
		local _newValue_1 = table.create(#points)
		local _callback_1 = function(p)
			return p:getY()
		end
		for _k, _v in points do
			_newValue_1[_k] = _callback_1(_v, _k - 1, points)
		end
		-- ▲ ReadonlyArray.map ▲
		local y = _newValue_1
		local maxX = math.max(unpack(x))
		local maxY = math.max(unpack(y))
		return Point2D.new(maxX, maxY)
	end
	function GRect:getMinPoint()
		local points = self:getPoints()
		-- ▼ ReadonlyArray.map ▼
		local _newValue = table.create(#points)
		local _callback = function(p)
			return p:getX()
		end
		for _k, _v in points do
			_newValue[_k] = _callback(_v, _k - 1, points)
		end
		-- ▲ ReadonlyArray.map ▲
		local x = _newValue
		-- ▼ ReadonlyArray.map ▼
		local _newValue_1 = table.create(#points)
		local _callback_1 = function(p)
			return p:getY()
		end
		for _k, _v in points do
			_newValue_1[_k] = _callback_1(_v, _k - 1, points)
		end
		-- ▲ ReadonlyArray.map ▲
		local y = _newValue_1
		local minX = math.min(unpack(x))
		local minY = math.min(unpack(y))
		return Point2D.new(minX, minY)
	end
	function GRect:getSides()
		return { Segment.new(self.p1, self.p2), Segment.new(self.p2, self.p3), Segment.new(self.p3, self.p4), Segment.new(self.p4, self.p1) }
	end
	function GRect:getPoints()
		return { self.p1, self.p2, self.p3, self.p4 }
	end
	function GRect:getSize()
		local length = GeneralMath.getDistance_2Points(self.p1, self.p2)
		local width = GeneralMath.getDistance_2Points(self.p2, self.p3)
		return Point2D.new(length, width)
	end
	function GRect:getArea()
		local size = self:getSize()
		return size:getX() * size:getY()
	end
	function GRect:getPerimeter()
		local size = self:getSize()
		return 2 * (size:getX() + size:getY())
	end
	function GRect:getRotation()
		local points = self:getPoints()
		local dy = points[2]:getY() - points[1]:getY()
		local dx = points[2]:getX() - points[1]:getX()
		local angle = math.atan2(dy, dx)
		return angle
	end
	function GRect:getAxisAlignedRect()
		local size = self:getSize()
		return GRect:createRectFromPoints(Point2D.new(size:getX() / 2, size:getY() / 2), Point2D.new(-size:getX() / 2, size:getY() / 2), Point2D.new(-size:getX() / 2, -size:getY() / 2), Point2D.new(size:getX() / 2, -size:getY() / 2))
	end
	function GRect:getBoundingBox()
		local points = self:getPoints()
		-- ▼ ReadonlyArray.map ▼
		local _newValue = table.create(#points)
		local _callback = function(p)
			return p:getX()
		end
		for _k, _v in points do
			_newValue[_k] = _callback(_v, _k - 1, points)
		end
		-- ▲ ReadonlyArray.map ▲
		local x = _newValue
		-- ▼ ReadonlyArray.map ▼
		local _newValue_1 = table.create(#points)
		local _callback_1 = function(p)
			return p:getY()
		end
		for _k, _v in points do
			_newValue_1[_k] = _callback_1(_v, _k - 1, points)
		end
		-- ▲ ReadonlyArray.map ▲
		local y = _newValue_1
		local minX = math.min(unpack(x))
		local maxX = math.max(unpack(x))
		local minY = math.min(unpack(y))
		local maxY = math.max(unpack(y))
		return GRect.new(Point2D.new(minX, minY), Point2D.new(maxX, minY), Point2D.new(maxX, maxY), Point2D.new(minX, maxY))
	end
	function GRect:isPointInside(point)
		local _binding = GeneralMath.rotatePoint(point:getX(), point:getY(), -self:getRotation())
		local x = _binding[1]
		local y = _binding[2]
		local rotatedPoint = Point2D.new(x, y)
		local axisAlignedRect = self:getAxisAlignedRect()
		local points = axisAlignedRect:getPoints()
		-- ▼ ReadonlyArray.map ▼
		local _newValue = table.create(#points)
		local _callback = function(p)
			return p:getX()
		end
		for _k, _v in points do
			_newValue[_k] = _callback(_v, _k - 1, points)
		end
		-- ▲ ReadonlyArray.map ▲
		local xValues = _newValue
		-- ▼ ReadonlyArray.map ▼
		local _newValue_1 = table.create(#points)
		local _callback_1 = function(p)
			return p:getY()
		end
		for _k, _v in points do
			_newValue_1[_k] = _callback_1(_v, _k - 1, points)
		end
		-- ▲ ReadonlyArray.map ▲
		local yValues = _newValue_1
		local minX = math.min(unpack(xValues))
		local maxX = math.max(unpack(xValues))
		local minY = math.min(unpack(yValues))
		local maxY = math.max(unpack(yValues))
		return rotatedPoint:getX() >= minX and rotatedPoint:getX() <= maxX and rotatedPoint:getY() >= minY and rotatedPoint:getY() <= maxY
	end
	function GRect:getCenter()
		return GeneralMath.centroid(self:getPoints())
	end
	function GRect:intersects(segment)
		local _exp = self:getSides()
		-- ▼ ReadonlyArray.some ▼
		local _result = false
		local _callback = function(side)
			return side:intersects(segment)
		end
		for _k, _v in _exp do
			if _callback(_v, _k - 1, _exp) then
				_result = true
				break
			end
		end
		-- ▲ ReadonlyArray.some ▲
		return _result
	end
	function GRect:intersectsRect(rect)
		local _exp = self:getSides()
		-- ▼ ReadonlyArray.some ▼
		local _result = false
		local _callback = function(side)
			return rect:intersects(side)
		end
		for _k, _v in _exp do
			if _callback(_v, _k - 1, _exp) then
				_result = true
				break
			end
		end
		-- ▲ ReadonlyArray.some ▲
		return _result
	end
	function GRect:getIntersection(segment)
		-- Loop through each side of the rectangle and check for intersection
		local _exp = self:getSides()
		-- ▼ ReadonlyArray.map ▼
		local _newValue = table.create(#_exp)
		local _callback = function(side)
			return side:getIntersection(segment)
		end
		for _k, _v in _exp do
			_newValue[_k] = _callback(_v, _k - 1, _exp)
		end
		-- ▲ ReadonlyArray.map ▲
		local intersections = _newValue
		local nonUndefinedIntersections = {}
		for _, intersection in intersections do
			if intersection ~= nil then
				table.insert(nonUndefinedIntersections, intersection)
			end
		end
		-- If the segment intersects, and the intersection is a segment, return the segment
		-- ▼ ReadonlyArray.filter ▼
		local _newValue_1 = {}
		local _callback_1 = function(intersection)
			return TS.instanceof(intersection, Segment)
		end
		local _length = 0
		for _k, _v in nonUndefinedIntersections do
			if _callback_1(_v, _k - 1, nonUndefinedIntersections) == true then
				_length += 1
				_newValue_1[_length] = _v
			end
		end
		-- ▲ ReadonlyArray.filter ▲
		local segmentIntersections = _newValue_1
		if #segmentIntersections > 0 then
			return segmentIntersections[1]
		end
		-- If the segment intersects, and the intersection is a point at the end of the segment, return the point
		-- ▼ ReadonlyArray.filter ▼
		local _newValue_2 = {}
		local _callback_2 = function(intersection)
			return TS.instanceof(intersection, Point2D)
		end
		local _length_1 = 0
		for _k, _v in nonUndefinedIntersections do
			if _callback_2(_v, _k - 1, nonUndefinedIntersections) == true then
				_length_1 += 1
				_newValue_2[_length_1] = _v
			end
		end
		-- ▲ ReadonlyArray.filter ▲
		local pointIntersections = _newValue_2
		-- If no intersections are found, return undefined
		-- Because if there are no segment intersections, there will always be at least one point intersection
		if #pointIntersections == 0 then
			return nil
		end
		-- If the point intersections is only one, return the point
		if #pointIntersections == 1 then
			return pointIntersections[1]
		end
		local _exp_1 = self:getPoints()
		-- ▼ ReadonlyArray.find ▼
		local _callback_3 = function(point)
			-- ▼ ReadonlyArray.some ▼
			local _result = false
			local _callback_4 = function(intersection)
				if point:equals(intersection) then
					return true
				end
			end
			for _k, _v in pointIntersections do
				if _callback_4(_v, _k - 1, pointIntersections) then
					_result = true
					break
				end
			end
			-- ▲ ReadonlyArray.some ▲
			return _result
		end
		local _result
		for _i, _v in _exp_1 do
			if _callback_3(_v, _i - 1, _exp_1) == true then
				_result = _v
				break
			end
		end
		-- ▲ ReadonlyArray.find ▲
		local intersectionAtRectCorner = _result
		if intersectionAtRectCorner ~= nil then
			return intersectionAtRectCorner
		end
		-- Else, find the next intersection, and return a segment from the first intersection to the next intersection
		local firstIntersection = pointIntersections[1]
		local secondIntersection = pointIntersections[2]
		return Segment.new(firstIntersection, secondIntersection)
	end
end
local default = GRect
return {
	default = default,
}
